Docker test kit

scubacabra opened the pull request:
Hi!

Just want to start off by saying thanks for putting this out here :+1:

It really helped getting some integration tests running with postgres!

I started off by simply adding a docker `PostgresService`.

I realized that since we use `specs2` that we couldn't use anything without a `testkit` supporting it.

I implemented that and got everything to work on our projects, and I thought that it might be nice to do a couple of extra things.
- create separate `testkit` projects that support `scalatest` and `specs2`
  - These projects also implement all the tests in their framework
- create a `core` project that the `testkit` projects depend on
- use `typesafe`'s config to replace all the hardcoded values in the traits
  - I thought that we might have a need to tweak values, and this has proved to be a good way to do that.

Let me know what you think?

scubacabra committed:
Run a `postgres` container via `docker`

scubacabra committed:
Support for tests using `specs2` framework

BeforeAfterAllStopOnError trait fails if there are any errors in the
`beforeAll` method(s).

This lets Docker start the images desired, because if docker isn't
up and the image(s) are also running, the tests aren't going to work.

scubacabra committed:
breaking up into multiple projects

- `core` -> `docker-it-scala-core`
  + has all the docker services

- `scalatest` -> `docker-testkit-scalatest`
  + `DockerTestKit` for `scalatest`
  + all tests that were originally in this repository

- `specs2` -> `docker-testkit-specs2`
  + `DockerTestKit` for `specs2`
  + all tests originall in this repo, but adapted to `specs2`

The `DockerTestKit` used to be in package `whisk.docker.test`, but with
this new multibuild, the package for both `scalatest` and `specs2` is
`whisk.docker`.

scubacabra committed:
Configuration via `typesafe` and `ficus`

All the original hardcoded values in the `Docker*Service` traits are
now in the `application.conf` in `core/test/resources`.

The new Typesafe config class has case classses to read the config and
transform that data into a `DockerContainer`.

These case classes have the defaults that were in the
`DockerReadyChecker`.

`DockerKit` loads the config and has a method that takes a path to
load.  `scalatest` and `specs2` tests run well.

One change in the config is that the `Kafka` service needs a `DOCKER_IP`
environmental variable that can be resolved in the typesafe config.

viktortnk commented:
Hi @jacobono . Thanks! That sounds very good.
I'll try to review and test it within next couple of days. If everything is good - happy to merge

scubacabra commented:
Cool @viktortnk 

My virtual box has a problem running the kafka container -- I could never get it to work before the changes.  Something about a lack of heap space as it is trying to start up.  The Zookeeper container starts, but the Kafka container just can't start up.

All of the tests work except for the Kafka test, and the AllAtOnce tests.

The only other change for the kafka tests to work is that an environmental variable of `DOCKER_IP` should be set for the `KAFKA_ADVERTISED_HOST_NAME` to be correct in `application.conf`.

Let me know if you have trouble with those two tests.

If everything looks good to you, would you have objections to publishing the binaries to jcenter/bintray?

viktortnk commented:
Yeah, I remember Kafka one was tricky. It was working in my OSX boot2docker setup, but didn't on linux machine with docker running locally.

Yes, will publish binaries

viktortnk commented:
@jacobono After a quick look I see that support for typesafeconfig-based containers is neat. And I'm happy to get it in. But I don't think it need to be part of the core.

In our systems we use dynamic configuration for containers in few places, so code-based definition will stay preferable for us.

Could you please move config-based container creation into separate module.

scubacabra commented:
I don't mind moving it -- I just want to explain how the typesafe
config implementation is the same as the code based implementation.

I tried to make it so that the current code base definition didn't
change after the typesafe config was added.  The config is translated
into the _same_ code based definition.

The code-based definition for `elasticsearchContainer` before

``` scala
trait DockerElasticsearchService extends DockerKit {

  val DefaultElasticsearchHttpPort = 9200
  val DefaultElasticsearchClientPort = 9300

  val elasticsearchContainer = DockerContainer("elasticsearch:1.7.1")
    .withPorts(DefaultElasticsearchHttpPort -> None, DefaultElasticsearchClientPort -> None)
    .withReadyChecker(
      DockerReadyChecker
        .HttpResponseCode(DefaultElasticsearchHttpPort, "/")
        .within(100 millis)
        .looped(20, 1250 millis)
    )

  abstract override def dockerContainers: List[DockerContainer] = elasticsearchContainer :: super.dockerContainers
}
```

ended up with a value of

``` scala
val elasticsearchContainer =
  DockerContainer(
    imageName = "elasticsearch:1.7.1",
    command: None,
    bindPorts = Map(DefaultElasticsearchHttpPort -> None, DefaultElasticsearchClientPort -> None),
    tty = false,
    stdinOpen: Boolean = false,
    links = Map.empty,
    env = Seq.empty,
    readyChecker = DockerReadyChecker
        .HttpResponseCode(DefaultElasticsearchHttpPort, "/")
        .within(100 millis)
        .looped(20, 1250 millis))
```

Is now expressed in a config file

``` scala
  elasticsearch {
    image-name = "elasticsearch:1.7.1"
    ready-checker {
      http-response-code {
        port = 9200
        path = "/"
        within = 100
        looped {
          attempts = 20
          delay = 1250
        }
      }
    }
    port-maps {
      default-elasticsearch-http-port {
        internal = 9200
      }
      default-elasticsearch-client-port {
        internal = 9300
      }
    }
  }
```

that is translated into the exact same end result value via
`DockerConfig.toDockerContainer`.

---

Does this show the equality of the two implementations?

If they are equal, will this implementation still mess up your dynamic
configurations in places?

---

I don't mind moving it to another module, but I'm having a hard time
seeing 2 things:
1. how to get a code-based definition (like before) and a typesafe config
   definition via `configureDockercontainer("docker.elasticsearch")` to
   work.
2. Currently `DockerTypesafeConfig` extends `DockerKit` for the implicit
   `docker` value and `ExecutionContext`.  And `DockerKit` imports
   `DockerTypesafeConfig` to read the configurations. I see a circular
   dependency if it is broken out.

Maybe you have some ideas on how to accomplish moving this around?

viktortnk commented:
@jacobono Services implementations here are mainly made to be just examples. Some of them we are using in our project as they were here, others we change. For example for ElasticSearch we generate unique clusterName. I don't see much reason for us using config + code to initialise container at the same time.

My assumption is that `DockerKit` might stay clear as it was before, without knowing anything about config-based definitions and operate with core entities like `DockerContainer`.

In a separate module, which would depend on `core`, you can define `ConfigBasedDockerKit` which will be extending `DockerKit` and adding method for `dockerConfig: Config` and `configureDockerContainer(name: String): DockerContainer`

That way if you want to use containers declared in config, you will be extending `ConfigBasedDockerKit` and we will keep going with our code-based definition from core.

Does it sound ok or am I missing something?

scubacabra commented:
That sounds great -- thanks for explaining.   I'll work on those changes.

scubacabra committed:
typesafe-config broken out into its own module

module-name: `docker-testkit-config`

Two ways to create a docker container, code-based definitions or
typesafe config.  Each is encapsulated in its own project and own
namespace.

The namespace for code-based definition is still `whisk.docker`, while
the new typesafe config namespace is `whisk.docker.config`.

All original code defined services are **also** in
`whisk.docker.config`.

modules `specs2` and `scalatest` only depend on `config` as a test
dependency -- all (src) compile dependencies are on `core`.

scubacabra committed:
README.md addition

scubacabra commented:
@viktortnk the config module is broken out like we discussed.

I also added a first shot at a `README`.

viktortnk commented:
@jacobono  Thanks. Yesterday I put configuration for publishing to bintray and made small README as well.

Could you please rebase your branch

scubacabra commented:
Rebased :+1:

viktortnk commented:
:+1:

viktortnk commented:
:+1: Merged. Thanks a lot for that!

viktortnk commented:
@jacobono Just published 0.2.1. Let me know if you have any issues using it as dependency

scubacabra commented:
@viktortnk Just used the `0.3.0` version and it works perfectly :+1: :grin: 

Thanks a bunch!
